juju/juju
# [dev] Run start hook on reboots

## Description of change
This PR enhances the uniter so that it can detect whether a reboot has occurred and notify
deployed charms by triggering an implicit start hook.

In order to detect reboots, the PR introduces the concept of a transient data
folder, a new agent configuration setting akin to "dataDir". On *nix targets we
get support for transient folders for free. All we need to do is to use a
tmpfs/ramdisk as the location of our transient data. By default (see `core/paths`),
juju will use `/var/run/juju` for storing transient data since that is guaranteed
to be a tmpfs volume (or a symlink to /run which is the equivalent for modern
systems). For windows, we need to use a low-level "syscall" to schedule the
automatic deletion of transient files when the system reboots (see extended
commit log for more details).

The `core/paths/transientfile` package offers a convenient abstraction for
creating transient files within the juju code-base. This functionality is
leveraged by yet another helper package `worker/common/reboot` which provides
per-entity (tag) one-off reboot notifications. The uniter uses this package
to queue start hooks if: a reboot is detected AND the unit has already been
started before (according to the uniter's state). Consequently, if we simply
restart the uniter agent, **no** start hook will be delivered. Similarly, if
we reboot the machine before the charm is installed the installation will
proceed as expected when the agent comes back online.

In addition, the deployer worker has been augmented with extra logic to clean up
any reboot-monitoring state once a unit gets removed. This is important to
avoid polluting manual machines (which may not be rebooted for a long time)
with junk state files.

Finally, to avoid triggering the start hook for existing units when upgrading
to 2.8, the PR includes an upgrade step that introspects the unit agents
running on the machine and pre-populates the reboot-handled state files
before the unit agents get restarted.

## QA steps

Deploy charms, trigger agent restarts and machine reboots and ensure that the
start hook fires when expected.

Or... Try the integration tests from this PR!

```
(cd tests ; ./main.sh hooks)
```

### Windows testing
Unfortunately, we don't have a reliable mechanism for running tests for windows charms. Given that the platform specific code for windows writes to the registry and that our CI machines are rarely restarted, none of the transient file tests are executed on windows. 

The plan (future) is to port the integration tests from this PR to run with windows workloads so we can catch any windows-specific regression.

In the meantime, I have manually verified in a windows VM that the code creates the transient file and that it gets correctly purged after a reboot so you will have to take my word for it :D

## Documentation changes
We need to document and communicate the change in the hook execution contracts.
